pkg/packetfilter/packetfilter.go

```diff
@@ -0,0 +1,273 @@
+	MarkValue string
+	Action    RuleAction
+	Proto     RuleProto
+	ClampType MssClampType
+}
+
+// Supported policy values are accept (which is the default) or drop.
+type ChainRegular struct {
+	Name   string
+	Policy ChainPolicy
+}
+
+type ChainIPHook struct {
+	Name     string
+	Type     ChainType
+	Hook     ChainHook
+	Priority ChainPriority
+	Policy   ChainPolicy
+	JumpRule *Rule
+}
+
+type Base interface {
+	// Chains
+	ChainExists(table TableType, chain string) (bool, error)
+	CreateIPHookChainIfNotExists(chain *ChainIPHook) error
+	CreateRegularChainIfNotExists(table TableType, chain *ChainRegular) error
+	DeleteIPHookChain(chain *ChainIPHook) error
+	DeleteRegularChain(table TableType, chain string) error
+	ClearChain(table TableType, chain string) error
+
+	// rules
+	Delete(table TableType, chain string, ruleSpec *Rule) error
+	InsertUnique(table TableType, chain string, position int, ruleSpec *Rule) error
+	PrependUnique(table TableType, chain string, ruleSpec *Rule) error
+	UpdateChainRules(table TableType, chain string, rules []*Rule) error
+	AppendUnique(table TableType, chain string, ruleSpec *Rule) error
+}
+
+type MssClampInterface interface {
+	SetMultiClusterMssClamp(clampType MssClampType, mssValue int) error
+	DelMultiClusterMssClamp() error
+	AddSubMultiClusterMssClamp(localCIDRs []string, destCIDRs []string) error
+	DelSubMultiClusterMssClamp(localCIDRs []string, destCIDRs []string) error
+}
+type GlobalNetEgressInterface interface {
+	CreateGlobalEgressForPods(key, snatCIDR, markValue string) (GlobalEgressHandle, error)
+	DeleteGlobalEgressForPods(handle GlobalEgressHandle) error
+	CreateGlobalEgressForNamespace(namespace, snatCIDR, markValue string) (GlobalEgressHandle, error)
+	DeleteGlobalEgressForNamespace(handle GlobalEgressHandle) error
+	AddPodIPToGlobalEgress(handle GlobalEgressHandle, podIP string) error
+	RemovePodIPToGlobalEgress(handle GlobalEgressHandle, podIP string) error
+}
+
+type Driver interface {
+	Base
+	MssClampInterface
+	GlobalNetEgressInterface
+}
+type DriverCreateFunc func() (Driver, error)
+
+type driverInfo struct {
+	createFunc DriverCreateFunc
+	name       string
+	isDefault  bool
+}
+
+var (
+	logger                 = log.Logger{Logger: logf.Log.WithName("PacketFilter")}
+	NewFunc                func() (Base, error)
+	NewMMssClampFunc       func() (MssClampInterface, error)
+	NewGlobalNetEgressFunc func() (GlobalNetEgressInterface, error)
+
+	drivers             []driverInfo
+	selectedDriverIndex int
+)
+
+func AddDriver(name string, isDefault bool, createFunc DriverCreateFunc) {
+	drivers = append(drivers,
+		driverInfo{
+			createFunc: createFunc,
+			name:       name,
+			isDefault:  isDefault,
+		})
+
+	logger.Info("%s registered as driver, isDefault:%t", name, isDefault)
+
+	if len(drivers) == 1 {
+		selectedDriverIndex = 0
+	} else {
+		for i := 0; i < len(drivers); i++ {
+			if drivers[i].isDefault {
+				selectedDriverIndex = i
+				break
+			}
+		}
+	}
+}
+
+type Adapter struct {
```

Comment:
So following the current `iptables` et al abstractions, this would be named `Interface` and would contain higher-level helper/utility methods that wrap one or more `Base` methods. This avoids implementations duplicating code. So following the [iptables](https://github.com/submariner-io/submariner/blob/628f134dae339c9cbd94ce23f2ea1177ab9cfdc9/pkg/iptables/iptables.go#L41) abstraction, `CreateIPHookChainIfNotExists`, `CreateRegularChainIfNotExists` et al would be in this interface.